ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/layered/LayeredPendingTransactions.java

```diff
@@ -314,55 +316,80 @@ public synchronized List<Transaction> getPriorityTransactions() {
-          logDiscardedTransaction(candidatePendingTx, selectionResult);
-        }
-
-        if (selectionResult.stop()) {
-          LOG.trace("Stopping selection");
-          break selection;
-        }
-
-        if (!selectionResult.selected()) {
-          // avoid processing other txs from this sender if this one is skipped
-          // since the following will not be selected due to the nonce gap
-          LOG.trace("Skipping remaining txs for sender {}", candidatePendingTx.getSender());
-          break;
+    for (final var entry : candidateTxsByScore.entrySet()) {
+      LOG.trace("Evaluating txs with score {}", entry.getKey());
+
+      for (final var senderTxs : entry.getValue()) {
+        LOG.trace("Evaluating sender txs {}", senderTxs);
+
+        if (!skipSenders.contains(senderTxs.sender())) {
+
+          for (final var candidatePendingTx : senderTxs.pendingTransactions()) {
+            final var selectionResult = selector.evaluateTransaction(candidatePendingTx);
+
+            LOG.atTrace()
+                .setMessage("Selection result {} for transaction {}")
+                .addArgument(selectionResult)
+                .addArgument(candidatePendingTx::toTraceLog)
+                .log();
+
+            if (selectionResult.discard()) {
+              invalidTransactions.add(candidatePendingTx);
+              logDiscardedTransaction(candidatePendingTx, selectionResult);
+            }
+
+            if (selectionResult.penalize()) {
+              penalizedTransactions.add(candidatePendingTx);
+              LOG.atTrace()
+                  .setMessage("Transaction {} penalized")
+                  .addArgument(candidatePendingTx::toTraceLog)
+                  .log();
+            }
+
+            if (selectionResult.stop()) {
+              LOG.trace("Stopping selection");
+              break selection;
+            }
+
+            if (!selectionResult.selected()) {
+              // avoid processing other txs from this sender if this one is skipped
+              // since the following will not be selected due to the nonce gap
+              LOG.trace("Skipping remaining txs for sender {}", candidatePendingTx.getSender());
+              skipSenders.add(candidatePendingTx.getSender());
+              break;
+            }
+          }
         }
       }
     }
 
     ethScheduler.scheduleTxWorkerTask(
-        () ->
-            invalidTransactions.forEach(
-                invalidTx -> {
-                  synchronized (this) {
-                    prioritizedTransactions.remove(invalidTx, INVALIDATED);
-                  }
-                }));
+        () -> {
+          invalidTransactions.forEach(
+              invalidTx -> {
+                synchronized (this) {
+                  prioritizedTransactions.remove(invalidTx, INVALIDATED);
+                }
+              });
+          penalizedTransactions.forEach(
+              penalizedTx -> {
+                synchronized (this) {
```

Comment:
If this is the preferred means to address the sync issue I outlined above, then I suppose that is fine. Might want a javadoc on that class indicating that it requires external syncronization.

Reply:
same as above, do you think the package javadoc is enough or we need also something here?

Reply:
I do think we also need something here, there are not many good paths that will get a confused devs to the package level javadoc.